applet/src/main/java/applet/EmeraldApplet.java

```diff
@@ -9,15 +9,83 @@ Team Emerald (in alphabetical order):
 
 import javacard.framework.APDU;
 import javacard.framework.Applet;
+import javacard.framework.ISO7816;
+import javacard.framework.ISOException;
 import javacard.framework.MultiSelectable;
 import javacard.framework.Util;
 
 public class EmeraldApplet extends Applet implements MultiSelectable {
-    public EmeraldApplet(byte[] buffer, short offset, byte length) {
+    public static final byte PIN_LENGTH = 4;
+
+    /**
+     * Persistent PIN storage in EEPROM.
+     * TODO `javacard.framework.OwnerPIN` is not suitable as we need to read PIN value for J-PAKE
+     */
+    private final byte[] pin; // TODO store pin securely
```

Comment:
Yep, that's a question, since it's not recommended to store PINs in primitive arrays. Maybe, only the hash or the key from some KDF is worth storing.

Reply:
Yes, this is not perfect. But it's the same for storing something deterministically derived from PIN.